bin/grouper-api

```diff
@@ -29,6 +31,66 @@ class Application(tornado.web.Application):
         super(Application, self).__init__(*args, **kwargs)
 
 
+class SendEmailThread(Thread):
+    """Background thread for sending emails from the async pool
+
+    This class thread will exist on multiple servers in a standard Grouper production environment
+    so we need to ensure that it's race-safe.
+    """
+
+    def send_emails(self, session, now_ts):
+        """Send emails due before now
+
+        This method finds and immediately sends any emails that have been scheduled to be sent
+        before the now_ts.
+
+        Args:
+            session (Session): Object for db session.
+            now_ts (datetime): The time to use as the cutoff (send emails before this point).
+
+        Returns:
+            int: Number of emails that were sent.
+        """
+        emails = session.query(AsyncNotification).filter(
+            AsyncNotification.sent == False,
+            AsyncNotification.send_after < now_ts,
+        ).all()
+        for email in emails:
+            # For atomicity, attempt to set the sent flag on this email to true if
+            # and only if it's still false.
+            # try:
+            update_ct = session.query(AsyncNotification).filter(
+                AsyncNotification.id == email.id,
+                AsyncNotification.sent == False
+            ).update({"sent": True})
+
+            # If it's 0, someone else won the race. Bail.
+            if update_ct == 0:
+                continue
+
+            try:
+                send_email_raw(settings, email.email, email.body)
+                email.sent = True
```

Comment:
Does this line or line 65 actually change the persisted state?  Should email.sent = False be set in the exception handler?